exasol_integration_test_docker_environment/lib/base/db_os_executor.py

```diff
@@ -0,0 +1,134 @@
+from abc import abstractmethod
+import fabric
+import docker
+from docker import DockerClient
+from typing import Optional, Protocol, runtime_checkable
+from docker.models.containers import Container, ExecResult
+from exasol_integration_test_docker_environment \
+    .lib.base.ssh_access import SshKey
+from exasol_integration_test_docker_environment \
+    .lib.data.database_info import DatabaseInfo
+from exasol_integration_test_docker_environment.lib.docker \
+    import ContextDockerClient
+
+
+class DockerClientFactory(Protocol):
+    """
+    Create a Docker client.
+    """
+    @abstractmethod
+    def client(self) -> DockerClient:
+        ...
+
+
+class ContextDockerClientFactory(DockerClientFactory):
+    def __init__(self, timeout: int = 100000):
+        self._timeout = timeout
+
+    def client(self) -> DockerClient:
+        return ContextDockerClient(timeout=self._timeout)
+
+
+@runtime_checkable
+class DbOsExecutor(Protocol):
+    """
+    This class provides an abstraction to execute commands inside a Docker
+    Container.  See concrete implementations in sub-classes
+    ``DockerExecutor`` and ``SshExecutor``.
+    """
+    @abstractmethod
+    def exec(self, cmd: str) -> ExecResult:
+        ...
+
+
+class DockerExecutor(DbOsExecutor):
+    def __init__(
+            self,
+            container_name: str,
+            timeout: int = 100000,
+            client_factory: Optional[DockerClientFactory] = None,
```

Comment:
why optional

Reply:
My initial idea was to make the productive usage as simple as possible.
Will be changed in next push.